homeassistant/components/synology_dsm/__init__.py

```diff
@@ -364,21 +423,47 @@ def _fetch_device_configuration(self):
         if self._with_upgrade:
             self.upgrade = self.dsm.upgrade
 
+        if self._with_system:
+            self.system = self.dsm.system
+
         if self._with_utilisation:
             self.utilisation = self.dsm.utilisation
 
         if self._with_surveillance_station:
             self.surveillance_station = self.dsm.surveillance_station
 
+    async def async_reboot(self):
+        """Reboot NAS."""
+        if not self.system:
+            _LOGGER.debug("async_reboot - self.system not ready: %s", self)
+            return
+        self._hass.async_add_executor_job(self.system.reboot)
+
+    async def async_shutdown(self):
+        """Shutdown NAS."""
+        if not self.system:
+            _LOGGER.debug("async_shutdown - self.system not ready: %s", self)
+            return
+        self._hass.async_add_executor_job(self.system.shutdown)
+
     async def async_unload(self):
         """Stop interacting with the NAS and prepare for removal from hass."""
         self._unsub_dispatcher()
 
     async def async_update(self, now=None):
         """Update function for updating API information."""
         self._async_setup_api_requests()
-        await self._hass.async_add_executor_job(self.dsm.update, self._with_information)
-        async_dispatcher_send(self._hass, self.signal_sensor_update)
+        try:
+            await self._hass.async_add_executor_job(
+                self.dsm.update, self._with_information
+            )
+            async_dispatcher_send(self._hass, self.signal_sensor_update)
+        except (SynologyDSMLoginFailedException, SynologyDSMRequestException) as err:
```

Comment:
Why adding `SynologyDSMLoginFailedException` here ?

I think `SynologyDSMRequestException` should be enough.

Reply:
same here
> because during boot (after reboot or scheduled downtime) DSM it self is already reachable (no connection error), but still login is unavailable (login failed) because DSM says "still booting, please try later"